### PR TITLE
INFR-50 - Netlify Content Security Policy

### PR DESCRIPTION
**Pull request by grahamplata**
Address content security policies

**Checklist:**
- [x] Covered by tests
- [ ] Ran it and it works as intended
- [x] Reviewed the diff before requesting a review
- [ ] Checked for unhandled edge cases - Share and Embed* but needs to be tested
- [x] [Linked INFR-50](https://linear.app/rilldata/issue/INFR-50/pentest-content-security-policy-implemented-in-an-insecure-manner) the issues it closes
- [x] Checked if the docs need to be updated. If so, create a separate Linear DOCS issue
- [ ] Intend to cherry-pick into the release branch
- [ ] I'm proud of this work!

### CODE REVIEW

File: netlify.toml [hunk 1]
```diff
@@ -1,23 +1,74 @@
...
+      connect-src 'self' https: wss://*.pusher.com; \
+      font-src 'self' https://fonts.gstatic.com https://*.usepylon.com;
       """
     Permissions-Policy = "geolocation=(),midi=(),sync-xhr=(self),microphone=(),camera=(),magnetometer=(),gyroscope=(),fullscreen=(self),payment=()"
     Referrer-Policy = "no-referrer"
     X-Content-Type-Options = "nosniff"
     Cross-Origin-Embedder-Policy = "credentialless"
     Cross-Origin-Opener-Policy = "unsafe-none"
     Cross-Origin-Resource-Policy = "cross-origin"
-    X-Frame-Options = "allow-from *"
```
Comment (himadrisingh) [inline netlify.toml:22]: we will need this for embedded dash

**Comment by grahamplata** [comment]
> ## Issues & Suggestions
> ### High Priority
> 1. **`/ingest` is a well-known PostHog proxy path — use something less obvious** (`web-admin/_redirects`, `posthog.ts`)
>    PostHog's docs recommend against using `/ingest` because ad blockers maintain lists of known proxy patterns. Suggest renaming to something like `/ph`:
>    
>    * `_redirects`: `/ph/* https://us.i.posthog.com/:splat 200`
>    * `posthog.ts`: `api_host: "/ph"`
> 2. **`connect-src 'self' https:` is overly broad** (`netlify.toml`, all three blocks)
>    `connect-src https:` allows fetch/XHR connections to _any_ HTTPS origin, which largely negates CSP's data exfiltration protection. Consider tightening to explicitly listed origins (e.g., `'self' https://*.rilldata.com https://*.usepylon.com wss://*.pusher.com`).
> 3. **`script-src` still has `'unsafe-inline'` and `'unsafe-eval'`** (`netlify.toml`, all three blocks)
>    These significantly weaken XSS protection. Likely required by current dependencies, but worth tracking as follow-up tech debt to adopt nonce-based or hash-based script loading.
> 4. **Netlify header precedence / `X-Frame-Options` conflict** (`netlify.toml`)
>    The default `/*` route sets `X-Frame-Options = "SAMEORIGIN"`. The embed and share routes intentionally allow framing (`frame-ancestors https:`) but don't set `X-Frame-Options`. If Netlify merges headers from the catch-all, `SAMEORIGIN` will conflict with `frame-ancestors https:`. Modern browsers prioritize CSP `frame-ancestors`, but older browsers may not. Consider explicitly omitting or overriding `X-Frame-Options` on the embed/share routes.
> 
> ### Medium Priority
> 5. **Significant duplication across three header blocks** (`netlify.toml`)
>    The only difference between the three CSP blocks is `frame-ancestors` and the presence of `X-Frame-Options`. All other headers are identical copy-paste. If one needs updating, all three must change — error-prone for future maintenance.
> 6. **`frame-ancestors https:` allows framing by any HTTPS site** (`netlify.toml` lines 38, 62)
>    For embed/share routes, any HTTPS site can embed your content. If there's a known set of embedding customers, restricting to specific domains would be more secure. If broad embedding is intentional, this is fine but worth documenting.
> 7. **One remaining unsanitized `{@html}` usage** (`TopK.svelte:143`)
>    `{@html ensureSpaces(percentage)}` — low risk since it operates on internal numeric data, but worth sanitizing for consistency and defense-in-depth.
> 
> ### Low Priority
> 8. **`style-src 'self' https: 'unsafe-inline'`** — the `https:` scheme allows loading stylesheets from any HTTPS origin. Consider restricting to known CDN/font origins if feasible.
> 9. **PR description is the placeholder** — "INSERT DESCRIPTION HERE" should be filled in. Checklist items "Covered by tests" and "Ran it and it works as intended" are unchecked, which is notable for a security-sensitive change.

I believe I addressed the lion's share but 

3. (script-src unsafe-inline/eval): Acknowledged as tech debt. These appear to be required by Svelte and other dependencies.
5. (Header duplication): Intentional tradeoff. Netlify doesn't support header inheritance, so keeping explicit blocks provides
   clarity and avoids unexpected merge behavior. The duplication is manageable given there are only three routes.
6. (frame-ancestors https:): Intentional. Broad embedding is a product requirement to support customers embedding dashboards
  on their own domains. Restricting to specific domains would require maintaining a customer allowlist and break self-serve
  embedding.

**Comment by grahamplata** [comment]
> Thanks for addressing the earlier feedback — `connect-src` tightening, `/ph` proxy path, DOMPurify on TopK, and `style-src` fix all look good.
> 
> Re your responses on the first review:
> 
> * **`unsafe-inline`/`unsafe-eval`**: Agree, fine as acknowledged tech debt if Svelte requires it.
> * **Header duplication**: Fair point that Netlify doesn't support inheritance. With the tightened `connect-src` though (8 explicit origins x3 blocks), the cost of a mistake is higher than before — see point 2 below.
> * **`frame-ancestors https:`**: Makes sense for self-serve embedding. No further concern there.
> 
> ### Remaining Issues
> 1. **`X-Frame-Options: ALLOWALL` is not a valid value** (`netlify.toml` lines 50, 75)
>    The spec only defines `DENY`, `SAMEORIGIN`, and the deprecated `ALLOW-FROM uri`. Browsers will ignore the unrecognized value, which happens to work here, but it may trigger security scanner warnings. Consider dropping `X-Frame-Options` from the default route entirely and relying solely on `frame-ancestors` across all routes.
> 2. **CSP duplication is now a high-priority concern** (`netlify.toml`)
>    The `connect-src` is now an explicit list of 8 origins, duplicated verbatim across all three header blocks. The only actual difference between the blocks is `frame-ancestors` and `X-Frame-Options`. If someone adds a new API dependency, they need to update three places — and getting one wrong means inconsistent security policies across routes. Consider extracting the shared headers into a `_headers` file or a build-time template.
> 3. **`https://us.posthog.com` in `connect-src` may be unnecessary** (`netlify.toml`, all three blocks)
>    The proxy goes through `/ph` (covered by `'self'`), so the browser shouldn't need direct connections to PostHog. The `ui_host` setting is only for PostHog toolbar UI links, not API calls. Removing it would tighten things further.
> 4. **PR description** is updated but "Covered by tests" and "Ran it and it works as intended" are still unchecked. For a security-sensitive change that touches CSP, embedding, and analytics proxying, it'd be good to confirm the embed and share routes actually work before merging (the checklist itself notes "Share and Embed* but needs to be tested").

  - Removed X-Frame-Options from all routes — now relying solely on frame-ancestors
  - Removed https://us.posthog.com from connect-src — proxied via /ph (covered by 'self')

Regarding duplication: Acknowledged. Netlify doesn't support header inheritance or templating. Given the CSP is now stable with 7 explicit origins, the tradeoff of 3 explicit blocks vs build-time generation complexity is acceptable.

**Comment by grahamplata** [comment]
Tested the header inheritance on the test environment (`ui.rilldata.in`) - it works as expected.

**Header Inheritance Verification:**

| Route | `frame-ancestors` | Shared Headers Inherited |
|-------|-------------------|-------------------------|
| `/` (root) | `'self'` | ✓ (defined here) |
| `/-/embed/*` | `https:` | ✓ |
| `/*/-/share/*` | `https:` | ✓ |

**All 7 shared headers are correctly inherited on embed/share routes:**
- `permissions-policy` ✓
- `referrer-policy` ✓
- `x-content-type-options` ✓
- `cross-origin-embedder-policy` ✓
- `cross-origin-opener-policy` ✓
- `cross-origin-resource-policy` ✓
- `strict-transport-security` ✓

Refactored `netlify.toml` to define shared headers once in `/*` and only override `Content-Security-Policy` in the embed/share routes. Added a comment explaining the header accumulation behavior.

**Review by ericpgreen2 — APPROVED** [review]
